listings/applications/advanced_factory/src/campaign/contributions.cairo

```diff
@@ -0,0 +1,106 @@
+use starknet::ContractAddress;
+
+#[starknet::interface]
+pub trait IContributable<TContractState> {
+    fn add(ref self: TContractState, contributor: ContractAddress, amount: u256);
+    fn get(self: @TContractState, contributor: ContractAddress) -> u256;
+    fn get_contributions_as_arr(self: @TContractState) -> Array<(ContractAddress, u256)>;
+    fn remove(ref self: TContractState, contributor: ContractAddress) -> u256;
+}
+
+#[starknet::component]
+pub mod contributable_component {
+    use core::array::ArrayTrait;
+    use starknet::{ContractAddress};
+    use core::num::traits::Zero;
+
+    #[storage]
+    struct Storage {
+        idx_to_contributor: LegacyMap<u32, ContractAddress>,
+        contributor_to_amt_idx: LegacyMap<ContractAddress, Option<(u256, u32)>>,
```

Comment:
Why do you need a `u32` index, isn't the contributor `ContractAddress` enough?
`LegacyMap<ContractAddress, u256>`, with no need for `Option` as the default value will be 0.

Edit: I suppose it's for iterating over the map element to construct an array. Maybe try to do it with a list of ContractAddress instead of a LegacyMap, and then you can iterate over it, get all the keys and construct your array. It should be slightly more storage efficient but benchmark is always a good idea.

Reply:
That's exactly the reason, I wanted to be able to iterate over the map. Just a note that I'm still experimenting with this storage design, I thought about using an array or list here and profiling the gas usage once I write enough tests to cover this.
Will push the most efficient storage to the PR and will post my benchmark findings.

Reply:
In a map, each insert and access will compute the hash of the key to get the address in memory, adding a slight total steps overhead. Also, you'll store each ContractAddress keys in storage + the index. Using List/Array also adds some overhead in terms of steps, but you still need to store all keys in storage (which make sense if your goal is to construct an array of tuples (key, value)).

I'm curious to see the result of the benchmark!

Reply:
Need some technical advice - when benchmarking, is the most important thing to measure the gas usage?
Is there anything else that needs to be measured?

Reply:
Yes gas is most important! There's not much to measure apart gas usage.

Reply:
@julio4 I benchmarked a couple of different solutions. Below I will list them best-to-worst; I will also link the branches that contain appropriate implementations, as well as post images of gas usage.

> Note: I renamed `contributor` to `pledger`, and correspondingly `contribution` to `pledge`.

I performed this benchmark by creating tests for each of the function that manages all of the pledges.

1. By far, the best performing storage design was to use `LegacyMap` for both indices and pledges.
https://github.com/misicnenad/StarknetByExample/blob/adv-factory/listings/applications/crowdfunding/src/campaign/pledgeable.cairo#L21
```rust
    struct Storage {
        index_to_pledger: LegacyMap<u32, ContractAddress>,
        pledger_to_amount: LegacyMap<ContractAddress, u256>,
        pledger_count: u32,
        total_amount: u256,
    }
```
![1st i_to_p legacymap and p_to_a legacymap png](https://github.com/NethermindEth/StarknetByExample/assets/19427053/bfb0accf-4476-48a7-89f9-c7792c0ddfeb)

2. The second best was to use Alexandria's `List` to index all of the pledgers:
https://github.com/misicnenad/StarknetByExample/blob/adv-fac-list/listings/applications/crowdfunding/src/campaign/pledgeable.cairo
```rust
    struct Storage {
        pledgers: List<ContractAddress>,
        pledger_to_amount: LegacyMap<ContractAddress, u256>,
        total_amount: u256,
    }
```
![2nd p List and p_to_a legacymap](https://github.com/NethermindEth/StarknetByExample/assets/19427053/d88d7281-9840-4b3a-9652-f38a69dad272)

3. The Storage design I originally used was 2x slower than previous two: 
https://github.com/misicnenad/StarknetByExample/blob/adv-fac-leg-map-with-index/listings/applications/crowdfunding/src/campaign/pledgeable.cairo
```rust
struct Storage {
        index_to_contributor: LegacyMap<u32, ContractAddress>,
        contributor_to_amount_index: LegacyMap<ContractAddress, Option<(u256, u32)>>,
        contributor_count: u32,
        total_amount: u256,
}
```
![3rd i_to_p legacymap and p_to_a_i legacymap](https://github.com/NethermindEth/StarknetByExample/assets/19427053/e23d0118-ded7-4183-b70c-ed434562488a)

---
P.S. I also tested things out with using just one `List` that contains both pledger addresses and pledged amounts, but this had the worst performance - when adding/removing a large number of pledgers, the test would run out of gas.
```rust
    struct Storage {
        pledges: List<(ContractAddress, u256)>,
        total_amount: u256,
    }
```
![image](https://github.com/NethermindEth/StarknetByExample/assets/19427053/ce34b735-274c-4184-8784-ee2935fecf40)